src/dra/63ED4.c

```diff
@@ -0,0 +1,367 @@
+// that this function should be the start of this new file.
+
+void func_80103ED4(void) {
+    char saveFile[32];
+    s32 memCardClose;
+    s32 i;
+    s32 case1_state;
+
+    switch (D_80137E4C) {
+    case 0:
+        MemcardInit();
+        D_80137E50 = 4;
+        D_80137E4C++;
+        return;
+    case 1:
+        // This really should be doable as a switch, but that doesn't match.
+        case1_state = func_800E9880(D_80097924, 0);
+        if (case1_state == 0) {
+            return;
+        }
+        if (case1_state == -1) {
+            if (--D_80137E50 == -1) {
+                D_80137E4C = 7;
+            }
+            return;
+        }
+        if (case1_state == -3) {
+            if (--D_80137E50 == -1) {
+                D_80137E4C = 8;
+            }
+            return;
+        }
+        if (case1_state == -2) {
+            D_80137E4C = 9;
+            return;
+        }
+        MemcardInit();
+        D_80137E4C++;
+        return;
+    case 2:
+        if (MemcardParse(D_80097924, 0) >= 0) {
+            D_80137E50 = 10;
+            if (D_8006C378 >= 0) {
+                i = 0;
+                if (D_80137E54 == 2) {
+                    for (i = 0; i < 15; i++) {
+                        MakeMemcardPath(saveFile, i);
+                        if (MemcardDetectSave(D_80097924, saveFile, 0) != 0) {
+                            break;
+                        }
+                    }
+                    if ((i == 15) &&
+                        (GetMemcardFreeBlockCount(D_80097924) == 0)) {
+                        D_80137E54 = 3;
+                    }
+                }
+                D_80137E4C += 2;
+                return;
+            } else {
+                D_80137E4C += 1;
+                return;
+            }
+        }
+        break;
+    case 3:
+        for (i = 0; i < 15; i++) {
+            MakeMemcardPath(saveFile, i);
+            if (MemcardDetectSave(D_80097924, saveFile, 0) == 0) {
+                break;
+            }
+        }
+        if (i != 15) {
+            if (GetMemcardFreeBlockCount(D_80097924) == 0) {
+                D_80137E4C = 10;
+                return;
+            }
+            D_8006C378 = i;
+            D_80137E4C++;
+            return;
+        }
+        D_80137E4C = 10;
+        return;
+    case 4:
+        MakeMemcardPath(saveFile, D_8006C378);
+        // careful with i here, it's not an iterator.
+        if (MemcardDetectSave(D_80097924, saveFile, 0) == 1) {
+            i = 0;
+        } else {
+            i = 1;
+            if (GetMemcardFreeBlockCount(D_80097924) == 0) {
+                D_80137E4C = 10;
+                return;
+            }
+        }
+        // I believe the rand() call here selects the icon on the save
+        // in the save-select screen.
+        StoreSaveData(g_Pix, D_8006C378, rand() & 0xF);
+        if (MemcardWriteFile(D_80097924, 0, saveFile, g_Pix, 1, i) != 0) {
```

Comment:
I'd suggest renaming i to shouldCreate or similar but I guess the variable gets reused. Does it still match if the i in case 4 is another var?
```
s32 MemcardWriteFile(
    s32 nPort, s32 nCard, const char* name, void* data, s32 flags, s32 create)
```